File: ts/11.binary-tree/findMaxDepth.test.ts
import { BinaryTree, BinaryTreeImpl, Node, NodeImpl } from './';
import { describe, beforeEach, test, assert } from 'vitest';
import { findMaxDepth } from './';

describe('binary tree: problem#1', () => {
    let binaryTree: BinaryTree<number>;
    beforeEach(() => {
        binaryTree = new BinaryTreeImpl<number>(new NodeImpl(1));
    });

    test('[best test case] find the max depth of the binary tree', () => {
        let root = binaryTree.getRoot() as Node<number>;
        // right
        root.pushRightLeaf(3);
        // left
        let generateNode = root.pushLeftLeaf(2);
        generateNode.pushLeftLeaf(4);
        generateNode = generateNode.pushRightLeaf(5);
        generateNode.pushRightLeaf(6);

        assert.equal(findMaxDepth({ ...root }), 4);
    });

    test('find the max depth of the binary tree with only the root node', () => {
        let root = binaryTree.getRoot() as Node<number>;
        assert.equal(findMaxDepth({ ...root }), 1);
    });

    test('[worst test case] find the max depth of the binary tree', () => {
        let root = binaryTree.getRoot() as Node<number>;
        root.pushRightLeaf(2)
            .pushRightLeaf(3)
            .pushRightLeaf(4)
            .pushRightLeaf(5)
            .pushRightLeaf(6);
        assert.equal(findMaxDepth({ ...root }), 6);
    });
});

File: golang/11.binary-tree/main/find_max_depth_test.go
package main

import (
	"github.com/AmirAhmadzadeh/problems/binary_tree"
	"testing"
)

func TestFindMaxDepth1(t *testing.T) {
	binaryTree := binary_tree.New(1)
	root := binaryTree.GetRoot()
	// right
	root.PushRightLeaf(2)
	// left
	generateNode := root.PushLeftLeaf(2)
	generateNode.PushLeftLeaf(4)
	generateNode = generateNode.PushRightLeaf(5)
	generateNode.PushRightLeaf(6)
	expectedResult := 4
	result := findMaxDepth(binaryTree)

	if expectedResult != result {
		t.Errorf("1.Expected (%v) is not same as"+
			" actual string (%v)", expectedResult, result)
	}

	binaryTree = binary_tree.New(1)
	expectedResult = 1
	result = findMaxDepth(binaryTree)
	if expectedResult != result {
		t.Errorf("2.Expected (%v) is not same as"+
			" actual string (%v)", expectedResult, result)
	}
}

// worst test case
func TestFindMaxDepth2(t *testing.T) {
	binaryTree := binary_tree.New(1)
	root := binaryTree.GetRoot()

	root.
		PushRightLeaf(2).
		PushRightLeaf(3).
		PushRightLeaf(4).
		PushRightLeaf(5).
		PushRightLeaf(6)

	expectedResult := 6
	result := findMaxDepth(binaryTree)

	if expectedResult != result {
		t.Errorf("1.Expected (%v) is not same as"+
			" actual string (%v)", expectedResult, result)
	}
}
